### PR TITLE
Back swipe hero

### LINKED ISSUE #9446: Create an app bar parallax transition effect for iOS
Native iOS navigations have a different transition for the body vs the app bar. While the body has a parallax + x axis and z axis slide, the app bar has a parallax cross fade.

### PR DESCRIPTION
**Pull request by xster**
Fixes #9446

Lets Heroes be initiated from a back swipe gesture. 
Changes signature for NavigatorObserver.didStartUserGesture from

```dart
void didStartUserGesture();
```

to

```dart
void didStartUserGesture(Route<dynamic> route, Route<dynamic> previousRoute);
```

Measures and inserts hero flight into overlay on first frame. 
Use in Cupertino nav bars.

giphy-8

Breaking signature announce: https://groups.google.com/forum/#!topic/flutter-dev/SBAlcp6W_3g

### CODE REVIEW

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 1]
```diff
@@ -410,7 +429,9 @@ class _HeroFlight {
       assert(type != null);
       switch (type) {
         case HeroFlightDirection.pop:
-          return initial.value == 1.0 && initial.status == AnimationStatus.reverse;
+          // Pops may be initiated by a user gesture that manually drives the
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: Was there a way to for initial.value == 1.0 and initial.status != AnimationStatus.reverse absent a user triggered transition? It looks like we've dropped that now.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 2]
```diff
@@ -532,14 +553,11 @@ class HeroController extends NavigatorObserver {
   /// linear [Tween<Rect>].
   HeroController({ this.createRectTween });
 
-  /// Used to create [RectTween]s that interpolate the position of heros in flight.
+  /// Used to create [RectTween]s that interpolate the position of heroes in flight.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: NICE

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 3]
```diff
@@ -559,45 +577,60 @@ class HeroController extends NavigatorObserver {
...
-  void didStartUserGesture() {
-    _questsEnabled = false;
-  }
-
-  @override
-  void didStopUserGesture() {
-    _questsEnabled = true;
+  void didStartUserGesture(Route<dynamic> route, Route<dynamic> previousRoute) {
+    assert(navigator != null);
+    assert(route != null);
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: It's OK if previousRoute is null?

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 4]
```diff
@@ -559,45 +577,60 @@ class HeroController extends NavigatorObserver {
...
 
   // If we're transitioning between different page routes, start a hero transition
   // after the toRoute has been laid out with its animation's value at 1.0.
-  void _maybeStartHeroTransition(Route<dynamic> fromRoute, Route<dynamic> toRoute, HeroFlightDirection flightType) {
-    if (_questsEnabled && toRoute != fromRoute && toRoute is PageRoute<dynamic> && fromRoute is PageRoute<dynamic>) {
+  void _maybeStartHeroTransition(
+    Route<dynamic> fromRoute,
+    Route<dynamic> toRoute,
+    HeroFlightDirection flightType,
+    [bool fromGesture = false]
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: I don't think this needs to be an optional parameter since it's private and only called in 2 places and you have to specify it in one of them.

The parameter name sounds like a gesture descriptor. Maybe isUserPopTransition.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 5]
```diff
@@ -559,45 +577,60 @@ class HeroController extends NavigatorObserver {
...
-      if (flightType == HeroFlightDirection.pop && animation.status == AnimationStatus.dismissed)
+      if (flightType == HeroFlightDirection.pop && animation.status == AnimationStatus.dismissed) {
         return;
+      }
 
-      // Putting a route offstage changes its animation value to 1.0. Once this
-      // frame completes, we'll know where the heroes in the `to` route are
-      // going to end up, and the `to` route will go back onstage.
-      to.offstage = to.animation.value == 0.0;
+      // For gesture driven pops into a page that maintainState, let the
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: For pop transitions driven by a user gesture: if the "to" page has maintainState = true, then the heroes final dimensions can be measured immediately because their page's layout is still valid.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 6]
```diff
@@ -559,45 +577,60 @@ class HeroController extends NavigatorObserver {
...
+      }
 
-      WidgetsBinding.instance.addPostFrameCallback((Duration value) {
-        _startHeroTransition(from, to, animation, flightType);
-      });
     }
   }
 
-  // Find the matching pairs of heros in from and to and either start or a new
+  // Find the matching pairs of heroes in from and to and either start or a new
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: NICE

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 7]
```diff
@@ -559,45 +577,60 @@ class HeroController extends NavigatorObserver {
...
   // hero flight, or divert an existing one.
   void _startHeroTransition(
     PageRoute<dynamic> from,
     PageRoute<dynamic> to,
     Animation<double> animation,
     HeroFlightDirection flightType,
+    bool fromGesture,
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: same comment as before about this parameter name

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 8]
```diff
@@ -348,11 +348,14 @@ class NavigatorObserver {
...
+  /// The [Navigator]'s route `route` is being moved by a user gesture.
   ///
-  /// For example, this is called when an iOS back gesture starts, and is used
-  /// to disabled hero animations during such interactions.
-  void didStartUserGesture() { }
+  /// For example, this is called when an iOS back gesture starts.
+  ///
+  /// If present, the route immediately below `route` is `previousRoute`.
+  /// Though the gesture may not necessarily conclude at `previousRoute` if
+  /// canceled.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/navigator.dart]: canceled => it is canceled.

Perhaps it's obvious but it seems like it would be worth explaining how a pop might be canceled and what that means vis didStopUserGesture().

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 9]
```diff
@@ -348,11 +348,14 @@ class NavigatorObserver {
   /// The [Navigator] replaced `oldRoute` with `newRoute`.
   void didReplace({ Route<dynamic> newRoute, Route<dynamic> oldRoute }) { }
 
-  /// The [Navigator]'s routes are being moved by a user gesture.
+  /// The [Navigator]'s route `route` is being moved by a user gesture.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/navigator.dart]: We should explain who needs to call this and when.  From the sound of it, it's just a notification.

The doc should link to and summarize the role of didStopUserGesture as well as Hero.transitionFromUserGestures.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 10]
```diff
@@ -1911,8 +1914,17 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   void didStartUserGesture() {
     _userGesturesInProgress += 1;
     if (_userGesturesInProgress == 1) {
+      final Route<dynamic> route = _history.last;
+      final Route<dynamic> previousRoute = route.willHandlePopInternally
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/navigator.dart]:
```
      final Route<dynamic> previousRoute = route.willHandlePopInternally || history.length < 2
        ? null
        : _history[_history.length - 2]
```

**Review by HansMuller — COMMENTED** [review]
I think this is a good API change. I had a few comments about the details.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 11]
```diff
@@ -123,8 +123,10 @@ class Hero extends StatefulWidget {
     this.createRectTween,
     this.flightShuttleBuilder,
     this.placeholderBuilder,
+    this.transitionForUserGestures = false,
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: Ah, this is sensible

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 12]
```diff
@@ -176,31 +178,48 @@ class Hero extends StatefulWidget {
   /// left in place once the Hero shuttle has taken flight.
   final TransitionBuilder placeholderBuilder;
 
+  /// Whether to perform the hero transition if the [PageRoute] transition was
+  /// triggered by a user gesture such as a back swipe on iOS.
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: Thanks

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 13]
```diff
@@ -176,31 +178,48 @@ class Hero extends StatefulWidget {
   /// left in place once the Hero shuttle has taken flight.
   final TransitionBuilder placeholderBuilder;
 
+  /// Whether to perform the hero transition if the [PageRoute] transition was
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: Yes. Though from this perspective (docs of Hero), the reader shouldn't really be concerned about the direction of the transition.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 14]
```diff
@@ -176,31 +178,48 @@ class Hero extends StatefulWidget {
...
+  /// triggered by a user gesture such as a back swipe on iOS.
+  ///
+  /// If [Hero]s with the same [tag] on both the from and the to routes have
+  /// [transitionForUserGestures] set to true, a back swipe gesture will
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: I agree. Though I'm not too sure what to put more specifically. 

I think didStartUserGesture is a bit of an implementation detail (in terms of how the HeroController receives events but we can envision changing that implementation detail without having to change the 'API' in the future).

Since I can't think of a specific other use case for the trigger and it's API stable now, it's probably better to be very specific in the doc and later expand than trying to restrict the API in the future.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 15]
```diff
@@ -176,31 +178,48 @@ class Hero extends StatefulWidget {
...
+  /// The route being popped to or the bottom route must also have
+  /// [PageRoute.maintainState] set to true for a gesture triggered hero
+  /// transition to work.
+  ///
+  /// Defaults to false and cannot be null.
+  final bool transitionForUserGestures;
+
   // Returns a map of all of the heroes in context, indexed by hero tag.
-  static Map<Object, _HeroState> _allHeroesFor(BuildContext context) {
+  static Map<Object, _HeroState> _allHeroesFor(BuildContext context, bool fromGesture) {
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: SG. Changing here and below.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 3 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: It is. And it should be ok for _maybeStartHeroTransition to drop it in that case too.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 4 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: Removed optional. Though I kept it the same name as above (to reduce cognitive load for future readers)

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 5 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: replaced with your wording

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 9 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/navigator.dart]: Cross referenced didStopUserGesture. 

Though this is a more generic callback API than heroes. We should probably not reference users of the API from the API.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 8 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/navigator.dart]: Added a bit more details.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 7 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: I'm not sure I understood. Can you rephrase what you meant?

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 10 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/navigator.dart]: Shortened.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/heroes.dart]: I have hoping to keep the gesture concept out of the flight and the manifests. Added the assert back.

**Review by HansMuller — APPROVED** [review]
LGTM
